### ISSUE TITLE
RealmChangeListener works incorrect when using findAllAsync 

### ISSUE THREAD

**Issue by ZherebtsovAlexandr** [realm/realm-java#2317]
This sample works correct, `RealmChangeListener.onChange` does't call when changed `Place`:

```java
RealmResults<Event> events = Realm.getDefaultInstance()
                .where(Event.class)
                .equalTo("eventId", 25518)
                .findAll();
events.addChangeListener(new RealmChangeListener() {
	@Override
	public void onChange() {
		Log.d("Realm", "Events was changed");
	}
});

Place place = Realm.getDefaultInstance()
                .where(Place.class)
                .findFirst();
Realm.getDefaultInstance().beginTransaction();
place.setTitle("Updated place");
Realm.getDefaultInstance().commitTransaction();
```

But if used async version `findAllAsync()` and `load()` then `onCnange()` will called even when changed `Place`:

```java
RealmResults<Event> events = Realm.getDefaultInstance()
                .where(Event.class)
                .equalTo("eventId", 25518)
                .findAllAsync();
events.load();                
events.addChangeListener(new RealmChangeListener() {
	@Override
	public void onChange() {
		Log.d("Realm", "Events was changed");
	}
});

Place place = Realm.getDefaultInstance()
                .where(Place.class)
                .findFirst();
Realm.getDefaultInstance().beginTransaction();
place.setTitle("Updated place");
Realm.getDefaultInstance().commitTransaction();
```

My `Realm` models:

```java
public class Place extends RealmObject {

    @PrimaryKey
    private int placeId;
    private String title;

    public int getPlaceId() {
        return placeId;
    }

    public void setPlaceId(int placeId) {
        this.placeId = placeId;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }
        
}
```

```java
public class Event extends RealmObject {

    @PrimaryKey
    private int eventId;
    private String title;

    public int getEventId() {
        return eventId;
    }

    public void setEventId(int eventId) {
        this.eventId = eventId;
    }
    
    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }
}
```

Realm version: `0.87.4`

**Comment by nhachicha** [comment]
Hi @ZherebtsovAlexandr 
`load` will block the current thread until the query returns, then it invoke any registered callback.

The first time the callback is notified we store an information `version` that allows us to avoid notifying the callbacks, if the underlying data used by the query didn't change. 

Since you're registering the callback after `load` finishes, this information is not stored yet (since you didn't have any callback yet). So the callback is triggered after you add a `Place` because we don't have an initial `version` to compare with.

Note: the registered `RealmChangeListener ` will not only trigger for change, but after the query completion. Imagine this use case

```java
 events.addChangeListener(new RealmChangeListener() {
    @Override
    public void onChange() {
        Log.d("Realm", "Events was changed");
    }
});
events.load();
```
Now the callback will trigger after the async query finishes (we do have a `version` to compare with now)

adding a new `Place` will not trigger the callback now (since it doesn't impact the initial query)
```
Place place = Realm.getDefaultInstance()
                .where(Place.class)
                .findFirst();
Realm.getDefaultInstance().beginTransaction();
place.setTitle("Updated place");
Realm.getDefaultInstance().commitTransaction(); 
```
_the above will not trigger the `RealmChangeListener` callback._

**Comment by ZherebtsovAlexandr** [comment]
Hi @nhachicha 

Thanks for the detailed answer and explanation. But I still need your comments. :)

If I understand you correctly, we are talking about this code:

**RealmResults.java**

```java
/**
     * Notifies all registered listeners.
     */
    void notifyChangeListeners() {
        if (listeners != null && !listeners.isEmpty()) {
            // table might be null (if the async query didn't complete
            // but we have already registered listeners for it)
            if (pendingQuery != null && !isCompleted) return;

            //FIXME: still waiting for Core to provide a fix
            //       for crash when calling _sync_if_needed on a cleared View.
            //       https://github.com/realm/realm-core/pull/1390
            long version;
            try {
                version = table.sync();
            } catch (DeletedRealmListException e) {
                // Although this RealmResults won't be updated anymore, it is good to give user a chance to do update.
                // When the onChange called this time, user can use isValid to check if the RealmList has been deleted.
                version = TABLE_VIEW_VERSION_REALM_LIST_DELETED;
                RealmLog.d("The parent RealmList has been deleted already.");
            }
            if (currentTableViewVersion != version) {
                currentTableViewVersion = version;
                for (RealmChangeListener listener : listeners) {
                    listener.onChange();
                }
            }

            // Since the parent RealmList has been removed, this RealmResults won't be updated anymore.
            // We just remove the change listeners from this to avoid unnecessary callings in the future.
            if (version == TABLE_VIEW_VERSION_REALM_LIST_DELETED) {
                listeners.clear();
            }
        }
    }
```
 I tried to debug the code:

**Option 1:**
Registering the callback before load  finishes.

Watches variables:

`currentTableViewVersion = -1 (TABLE_VIEW_VERSION_NONE)`
`version = 2`

`(currentTableViewVersion != version)` is true

**Option 2:**
Registering the callback after load  finishes.

Watches variables:

`currentTableViewVersion = 0`
`version = 2`

`(currentTableViewVersion != version)` is true

So, in any case, `RealmChangeListener` will call when change `Place`. Am I missing something?

Thank you!

**Comment by nhachicha** [comment]
Hi @ZherebtsovAlexandr 
yes I was referring to the above code.

In **Option1:**
T0: The async query completes
       we have one callback registered we set the initial value of `currentTableViewVersion` and we notify the callback
T1: a new commit happens, we rerun the query   
      `currentTableViewVersion`  is still the same we don't notify the listener

In **Option2:**
T0: The async query completes
      no callback registered so no update of `currentTableViewVersion` still `== -1` 
T1: a new commit happens, we rerun the query
      a callback is registered we update `currentTableViewVersion` 
T2: a new commit to `Place` will not notify our callback since we have already a version to compare against

Bottom line, yes the callback is triggered for both Options, but for different reasons, plus once you get a valid `currentTableViewVersion` to compare against, you have the expected behaviour (i.e only changes to the concerned Table will trigger a notification)

like I said before this is not fine grained notification #989 this is just a hack to reduce the number of irrelevant callbacks attached to your `RealmResults` or `RealmObject` 

Cheers

**Comment by beeender** [comment]
@ZherebtsovAlexandr  I hope your issue is solved. To make it easier for us to tracking the issue, i am closing this now. Feel free to reopen it if needed. Thanks!
